compiler/noirc_evaluator/src/ssa/opt/constant_folding/mod.rs

```diff
@@ -134,10 +160,39 @@ fn constant_folding_post_check(context: &Context, dfg: &DataFlowGraph) {
     );
 }
 
+/// Find all blocks in a set of blocks in the CFG which are not dominated by any block other than themselves.
+fn collect_blocks_not_dominated_by_others(
+    dom: &mut DominatorTree,
+    blocks: &BTreeSet<BasicBlockId>,
+) -> Vec<BasicBlockId> {
+    // Equivalent to following, but avoids the O(n*n*n) complexity coming from the fact that `dominates` may walk up the tree:
+    // blocks.filter(|b| blocks.all(|a| *a == **b || !dom.dominates(*a, **b)))
+    let mut has_dominator = HashSet::new();
+    for block in blocks {
+        let mut dominator = *block;
+        let mut path = vec![dominator];
+        while let Some(next) = dom.immediate_dominator(dominator) {
+            if has_dominator.contains(&next) || blocks.contains(&next) {
+                has_dominator.extend(path);
+                break;
+            }
+            path.push(next);
+            dominator = next;
+        }
+    }
+    blocks.iter().filter(|b| !has_dominator.contains(b)).copied().collect()
```

Comment:
We could avoid this second iteration of the blocks. Could build a second result set that we insert into if we exit the while look without getting into `has_dominator.contains(&next) || blocks.contains(&next)`